posthog/tasks/verify_persons_data_in_sync.py

```diff
@@ -0,0 +1,147 @@
+"""
+
+
+@app.task(max_retries=1, ignore_result=True)
+def verify_persons_data_in_sync() -> None:
+    # :KLUDGE: Rather than filter on created_at directly which is unindexed, we look up the latest value in 'id' column
+    #   and leverage that to narrow down filtering in an index-efficient way
+    max_pk = Person.objects.filter(created_at__lte=now() - PERIOD_START).latest("id").id
+    person_data = list(
+        Person.objects.filter(
+            pk__lte=max_pk, pk__gte=max_pk - LIMIT * 5, created_at__gte=now() - PERIOD_END
+        ).values_list("id", "uuid", "team_id")[:LIMIT]
+    )
+    person_data.sort(key=lambda row: row[2])  # keep persons from same team together
+
+    results = Counter(
+        {
+            "total": 0,
+            "missing_in_clickhouse": 0,
+            "version_mismatch": 0,
+            "properties_mismatch": 0,
+            "distinct_ids_mismatch": 0,
+        }
+    )
+    for i in range(0, len(person_data), BATCH_SIZE):
+        batch = person_data[i : i + BATCH_SIZE]
+        results += _team_integrity_statistics(batch)
+    _emit_metrics(results)
+
+
+def _team_integrity_statistics(person_data: List[Any]) -> Counter:
+    person_ids = [id for id, _, _ in person_data]
+    person_uuids = [uuid for _, uuid, _ in person_data]
+    team_ids = list(set(team_id for _, _, team_id in person_data))
+
+    # :TRICKY: To speed up processing, we fetch all models in batch at once and store results in dictionary indexed by person uuid
+    pg_persons = _index_by(
+        list(
+            Person.objects.filter(id__in=person_ids).prefetch_related(
+                Prefetch("persondistinctid_set", to_attr="distinct_ids_cache")
+            )
+        ),
+        lambda p: p.uuid,
+    )
+
+    ch_persons = _index_by(
+        sync_execute(GET_PERSON_CH_QUERY, {"person_ids": person_uuids, "team_ids": team_ids}), lambda row: row[0]
+    )
+
+    ch_distinct_ids_mapping = _index_by(
+        sync_execute(GET_DISTINCT_IDS_CH_QUERY, {"person_ids": person_uuids, "team_ids": team_ids}),
+        lambda row: row[1],
+        flat=False,
+    )
+
+    result: Counter = Counter()
+    for pk, uuid, team_id in person_data:
+        # Person was deleted in the middle of processing, can ignore
+        if uuid not in pg_persons:
+            continue
+        result["total"] += 1
+        pg_person = pg_persons[uuid]
+        if uuid not in ch_persons:
+            result["missing_in_clickhouse"] += 1
+            continue
+        _, ch_version, ch_properties = ch_persons[uuid]
+        ch_properties = json.loads(ch_properties)
+        if ch_version != pg_person.version:
+            result["version_mismatch"] += 1
+            logger.info(
+                "Found version mismatch",
+                team_id=team_id,
+                uuid=uuid,
+                properties=pg_person.properties,
+                ch_properties=ch_properties,
+            )
+        if pg_person.properties != ch_properties:
```

Comment:
If version mismatches this is pretty much guaranteed to mismatch, so let's ignore this when versions match so we have some useful data from here.

Reply:
Let's not build business logic into these metrics and keep it KISS/as users would observe. Instead if we manage to solve the right problem _and_ they're indeed correlated we will see it from the graphs.

Reply:
I want to be able to tell if any other fields mismatch, when versions match, if you think this one is important to keep too we can have both, but I need metrics for `data doesn't match for properties or created_at, but versions match` <= as that's something I want to know immediately and find the problem in the code and fix.